File: benchmark/rclpy/publisher-endurance-test.py
#!/usr/bin/env python3
# Copyright (c) 2017 Intel Corporation. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
from builtin_interfaces.msg import *
import math
import rclpy
from sensor_msgs.msg import *
from std_msgs.msg import *
import threading
from time import time

def main():
  parser = argparse.ArgumentParser()
  parser.add_argument("-p", "--period", type=int, help="The period(ms) to send a topic")
  parser.add_argument("-r", "--run", type=int, help="How many times to run")
  args = parser.parse_args()
  if args.period is None:
    args.period = 1
  if args.run is None:
    args.run = 1

  rclpy.init()
  period = args.period / 1000
  print('The publisher will publish a JointState topic %s times every %sms.' % (args.run, args.period))
  start = time()
  node = rclpy.create_node('endurance_publisher_rclpy')
  publisher = node.create_publisher(JointState, 'endurance_topic')
  stamp = Time()
  stamp.sec = 123456
  stamp.nanosec = 789
  header = Header()
  header.stamp = stamp
  header.frame_id = 'main_frame'
  msg = JointState()
  msg.header = header
  msg.name = ['Tom', 'Jerry']
  msg.position = [1.0, 2.0]
  msg.velocity = [2.0, 3.0]
  msg.effort = [4.0, 5.0, 6.0]
  total_times = args.run
  sent_times = 0

  def publish_topic():
    nonlocal publisher
    nonlocal msg
    nonlocal timer
    nonlocal sent_times
    if sent_times > total_times:
      timer.cancel()
      node.destroy_node()
      rclpy.shutdown()
      diff = time() - start
      milliseconds, seconds = math.modf(diff)
      print('Benchmark took %d seconds and %d milliseconds.' % (seconds, round(milliseconds * 1000)))
    else:
      publisher.publish(msg)
      sent_times += 1
      timer = threading.Timer(period, publish_topic)
      timer.start()

  timer = threading.Timer(period, publish_topic)
  timer.start()

if __name__ == '__main__':
  main()

File: benchmark/rclnodejs/publisher-endurance-test.js
// Copyright (c) 2018 Intel Corporation. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

/* eslint-disable camelcase */
const app = require('commander');
const rclnodejs = require('../../index.js');

app
  .option('-p, --period [period_ms]', 'The period(ms) to send a topic')
  .option('-r, --run <n>', 'How many times to run')
  .parse(process.argv);

rclnodejs.init().then(() => {
  const JointState = 'sensor_msgs/msg/JointState';
  const state = {
    header: {
      stamp: {
        sec: 123456,
        nanosec: 789,
      },
      frame_id: 'main_frame',
    },
    name: ['Tom', 'Jerry'],
    position: [1, 2],
    velocity: [2, 3],
    effort: [4, 5, 6],
  };

  const time = process.hrtime();
  let node = rclnodejs.createNode('endurance_publisher_rclnodejs');
  let publisher = node.createPublisher(JointState, 'endurance_topic');
  let sentTimes = 0;
  let totalTimes = app.run || 1;
  let period = app.period || 1;
  console.log(`The publisher will publish a JointState topic ${totalTimes} times every ${period}ms.`);

  let timer = setInterval(() => {
    if (sentTimes++ > totalTimes) {
      rclnodejs.shutdown();
      clearInterval(timer);
      const diff = process.hrtime(time);
      console.log(`Benchmark took ${diff[0]} seconds and ${Math.ceil(diff[1] / 1000000)} milliseconds.`);
    } else {
      publisher.publish(state);
    }}, period);
  rclnodejs.spin(node);
}).catch((err) => {
  console.log(err);
});
